<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.6: http://docutils.sourceforge.net/" />
<title>LAST Tutorial</title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 5951 2009-05-18 18:03:10Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left{
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: left }

/* div.align-center * { */
/*   text-align: left } */

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block {
  margin-left: 2em ;
  margin-right: 2em }

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
<style type="text/css">

/* Style sheet for LAST HTML documents */
h1 { color: navy }
h2 { color: teal }
div.document { margin-left: auto; margin-right: auto; max-width: 45em }
strong { color: red }
.option-list td { padding-bottom: 1em }
table.field-list { border: thin solid green }

</style>
</head>
<body>
<div class="document" id="last-tutorial">
<h1 class="title">LAST Tutorial</h1>

<p>LAST finds similar regions between sequences, and aligns them.</p>
<div class="section" id="example-1-compare-the-human-and-fugu-mitochondrial-genomes">
<h2>Example 1: Compare the human and fugu mitochondrial genomes</h2>
<p>For our first example, we wish to find and align similar regions
between the human and fugu mitochondrial genomes.  You can find these
sequences in the examples directory: humanMito.fa and fuguMito.fa.  We
can compare them like this:</p>
<pre class="literal-block">
lastdb -cR01 humdb humanMito.fa
lastal humdb fuguMito.fa &gt; myalns.maf
</pre>
<p>The lastdb command creates several files whose names begin with
&quot;humdb&quot;.  The lastal command then compares fuguMito.fa to the humdb
files, and writes the alignments to a file called &quot;myalns.maf&quot;.</p>
<p>The &quot;-cR01&quot; option suppresses alignments caused by simple sequence
such as cacacacacacacacacacacaca.</p>
</div>
<div class="section" id="understanding-the-output">
<h2>Understanding the output</h2>
<p>The output has very long lines, so you need to view it without
line-wrapping.  For example, with a Unix/Linux/MacOsX command line,
you can use:</p>
<pre class="literal-block">
less -S myalns.maf
</pre>
<p>Each alignment looks like this:</p>
<pre class="literal-block">
a score=27 EG2=4.7e+04 E=2.6e-05
s humanMito 2170 145 + 16571 AGTAGGCCTAAAAGCAGCCACCAATTAAGAAAGCGTT...
s fuguMito  1648 142 + 16447 AGTAGGCTTAGAAGCAGCCACCA--CAAGAAAGCGTT...
</pre>
<p>The score is a measure of how strong the similarity is.  EG2 and E are
explained at <a class="reference external" href="last-evalues.html">last-evalues.html</a>.  Lines starting with &quot;s&quot; contain:
the sequence name, the start coordinate of the alignment, the number
of bases spanned by the alignment, the strand, the sequence length,
and the aligned bases.</p>
<p>The start coordinates are zero-based.  This means that, if the
alignment begins right at the start of a sequence, the coordinate is
0.  If the strand is &quot;-&quot;, the start coordinate is in the reverse
strand.</p>
<p>This alignment format is called <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format5">MAF (multiple alignment format)</a>.  You can
convert it to several other formats using <a class="reference external" href="maf-convert.html">maf-convert</a>.  You can make lastal produce a few other formats
with option -f (see <a class="reference external" href="lastal.html">lastal.html</a>).</p>
</div>
<div class="section" id="example-2-compare-vertebrate-proteins-to-invertebrate-proteins">
<h2>Example 2: Compare vertebrate proteins to invertebrate proteins</h2>
<p>Use the lastdb -p option to indicate that the sequences are proteins:</p>
<pre class="literal-block">
lastdb -p -cR01 invdb invertebrate.fa
lastal invdb vertebrate.fa
</pre>
</div>
<div class="section" id="example-3-compare-dna-sequences-to-protein-sequences">
<h2>Example 3: Compare DNA sequences to protein sequences</h2>
<p>Here we use the -F15 option, to specify translated alignment with a
score penalty of 15 for frameshifts:</p>
<pre class="literal-block">
lastdb -p -cR01 protdb proteins.fa
lastal -F15 protdb dnas.fa
</pre>
</div>
<div class="section" id="example-4-find-short-protein-alignments">
<h2>Example 4: Find short protein alignments</h2>
<p>LAST uses a <a class="reference external" href="last-matrices.html">scoring scheme</a> to find
similarities.  Some scoring schemes are good for long-and-weak
similarities, others for short-and-strong similarities.  If we seek
very short similarities, weak ones are hopeless (statistically
insignificant), so we had better focus on strong ones.  The PAM30
scoring scheme may work well:</p>
<pre class="literal-block">
lastdb -p -cR01 invdb invertebrate.fa
lastal -pPAM30 invdb vertebrate.fa
</pre>
<p>(How short is &quot;very short&quot;?  It depends on the amount of sequence data
we are searching, but perhaps roughly less than 40 amino acids.)</p>
</div>
<div class="section" id="example-5-align-human-dna-reads-to-the-human-genome">
<h2>Example 5: Align human DNA reads to the human genome</h2>
<p>Suppose we have DNA reads in a file called reads.fastq, in
fastq-sanger format.  We can align them to the human genome like
this:</p>
<pre class="literal-block">
lastdb -uNEAR humandb human/chr*.fa
lastal -Q1 -e120 humandb reads.fastq | last-split &gt; myalns.maf
</pre>
<p>This will use about 15 gigabytes of memory.</p>
<ul>
<li><p class="first">-uNEAR selects a <a class="reference external" href="last-seeds.html">seeding scheme</a> that makes it
better at finding short-and-strong similarities.  (It also changes
the default scoring scheme.)</p>
</li>
<li><p class="first">The -Q1 option indicates that the reads are in fastq-sanger format.</p>
</li>
<li><p class="first">The -e120 option requests alignments with score ≥ 120.  This is
intentionally a somewhat low score (high E-value): last-split then
discards low-confidence alignments, but it uses them to estimate the
ambiguity of high-confidence alignments.</p>
</li>
<li><p class="first">last-split reads the alignments produced by lastal, and looks for a
unique best alignment for each part of each read.  It allows
different parts of one read to match different parts of the genome.
It has several useful options, please see <a class="reference external" href="last-split.html">last-split.html</a>.</p>
</li>
</ul>
</div>
<div class="section" id="paired-reads">
<h2>Paired reads</h2>
<p>If you have paired reads, there are two options:</p>
<ol class="arabic">
<li><p class="first">Use last-pair-probs (see <a class="reference external" href="last-pair-probs.html">last-pair-probs.html</a>).</p>
</li>
<li><p class="first">Ignore the pairing information, and align the reads individually
(using last-split as above).  This may be useful because
last-pair-probs does not currently allow different parts of one
read to match different parts of the genome, though it does allow
the two reads in a pair to match (e.g.) different chromosomes.</p>
</li>
</ol>
</div>
<div class="section" id="fastq-format-confusion">
<h2>Fastq format confusion</h2>
<p>Unfortunately, there is more than one fastq format (see
<a class="reference external" href="http://nar.oxfordjournals.org/content/38/6/1767.long">http://nar.oxfordjournals.org/content/38/6/1767.long</a>).  Recently
(2013) fastq-sanger seems to be dominant, but if you have another
variant you need to change the -Q option (see <a class="reference external" href="lastal.html">lastal.html</a>).</p>
</div>
<div class="section" id="example-6-align-human-fasta-reads-to-the-human-genome">
<h2>Example 6: Align human fasta reads to the human genome</h2>
<p>If our reads are in fasta instead of fastq format, we simply omit -Q:</p>
<pre class="literal-block">
lastdb -uNEAR humandb human/chr*.fa
lastal -e120 humandb reads.fa | last-split &gt; myalns.maf
</pre>
<p>(In older versions of LAST, we had to set a short-and-strong scoring
scheme, but this is now done automatically by -uNEAR.)</p>
</div>
<div class="section" id="example-7-align-aardvark-fastq-reads-to-the-human-genome">
<h2>Example 7: Align aardvark fastq reads to the human genome</h2>
<p>In this case we expect weak similarities, so we omit -uNEAR.  We also
need to change the scoring scheme (because with -Q1 it defaults to a
short-and-strong scoring scheme):</p>
<pre class="literal-block">
lastdb -cR01 humandb human/chr*.fa
lastal -Q1 -r5 -q5 -a35 -b5 humandb reads.fastq &gt; myalns.maf
</pre>
<p>Option -r5 sets the match score to 5, -q5 sets the mismatch cost to 5,
while -a35 and -b5 set the gap cost to 35 + 5×(gap length).</p>
<p>(Why use 5:5:35:5 rather than 1:1:7:1?  The reason is that 5:5:35:5
has roughly the same scale as the fastq quality scores.  lastal uses
the quality scores to modify the alignment scores, and then rounds the
modified scores to integers.  If we used 1:1:7:1, the integer-rounding
would lose information.)</p>
</div>
<div class="section" id="very-short-reads">
<h2>Very short reads</h2>
<p>WARNING!  The standard score parameters do not align very short reads.
This is because the match score is 6 and the score threshold is 120,
so at least 20 high-quality matches are required (or a greater number
of low-quality matches).  In addition, last-split discards
low-confidence alignments.  To align very short reads, reduce lastal's
score threshold (-e) or increase last-split's error threshold (-m).</p>
<p>If the score threshold is too low, you will get meaningless, random
alignments.</p>
</div>
<div class="section" id="tuning-speed-sensitivity-memory-and-disk-usage">
<h2>Tuning speed, sensitivity, memory and disk usage</h2>
<ul>
<li><p class="first">You can make LAST faster by <a class="reference external" href="last-parallel.html">using multiple CPUs</a>.</p>
</li>
<li><p class="first">You can <a class="reference external" href="last-tuning.html">trade off speed, sensitivity, memory and disk usage</a>.</p>
</li>
</ul>
</div>
<div class="section" id="example-8-compare-the-cat-and-rat-genomes">
<h2>Example 8: Compare the cat and rat genomes</h2>
<p>If you have ~50 GB of memory and don't mind waiting a few days, this
is a good way to compare such genomes:</p>
<pre class="literal-block">
lastdb -cR11 -uMAM8 catdb cat.fa
lastal -m100 -E0.05 catdb rat.fa | last-split -m1 &gt; out.maf
</pre>
<p>This looks for a unique best alignment for each part of each rat
chromosome.  Omitting -m100 makes it faster but somewhat less
sensitive.  Omitting -uMAM8 reduces the memory use to ~10 GB and makes
it faster but considerably less sensitive.</p>
<p>This recipe aligns each rat base-pair to at most one cat base-pair,
but not necessarily vice-versa.  You can get strictly 1-to-1
alignments by swapping the sequences and running last-split again:</p>
<pre class="literal-block">
maf-swap out.maf | last-split -m1 &gt; out2.maf
</pre>
</div>
<div class="section" id="example-9-compare-the-human-and-chimp-genomes">
<h2>Example 9: Compare the human and chimp genomes</h2>
<p>For strongly similar genomes (e.g. 99% identity), something like this
is more appropriate:</p>
<pre class="literal-block">
lastdb -cR11 -uNEAR human human.fa
lastal -m50 -E0.05 human chimp.fa | last-split -m1 &gt; out.maf
</pre>
</div>
<div class="section" id="example-10-ambiguity-of-alignment-columns">
<h2>Example 10: Ambiguity of alignment columns</h2>
<p>Consider this alignment:</p>
<pre class="literal-block">
TGAAGTTAAAGGTATATGAATTCCAATTCTTAACCCCCCTATTAAACGAATATCTTG
|||||||| ||||||  |  ||  | |  |    || ||||||   |||||||||||
TGAAGTTAGAGGTAT--GGTTTTGAGTAGT----CCTCCTATTTTTCGAATATCTTG
</pre>
<p>The middle section has such weak similarity that its precise alignment
cannot be confidently inferred.</p>
<p>It is sometimes useful to estimate the ambiguity of each column in an
alignment.  We can do that using lastal option -j4:</p>
<pre class="literal-block">
lastdb -cR01 humdb humanMito.fa
lastal -j4 humdb fuguMito.fa &gt; myalns.maf
</pre>
<p>The output looks like this:</p>
<pre class="literal-block">
a score=17 EG2=9.3e+09 E=5e-06
s seqX 0 57 + 57 TGAAGTTAAAGGTATATGAATTCCAATTCTTAACCCCCCTATTAAACGAATATCTTG
s seqY 0 51 + 51 TGAAGTTAGAGGTAT--GGTTTTGAGTAGT----CCTCCTATTTTTCGAATATCTTG
p                %*.14442011.(%##&quot;%$$$$###&quot;&quot;!!!&quot;&quot;&quot;&quot;&amp;'(*,340.,,.~~~~~~~~~~~
</pre>
<p>The &quot;p&quot; line indicates the probability that each column is wrongly
aligned, using a compact code (the same as fastq-sanger format):</p>
<blockquote>
<table border="1" class="docutils">
<colgroup>
<col width="13%" />
<col width="37%" />
<col width="13%" />
<col width="37%" />
</colgroup>
<tbody valign="top">
<tr><td>Symbol</td>
<td>Error probability</td>
<td>Symbol</td>
<td>Error probability</td>
</tr>
<tr><td><tt class="docutils literal">!</tt></td>
<td>0.79 -- 1</td>
<td><tt class="docutils literal">0</tt></td>
<td>0.025 -- 0.032</td>
</tr>
<tr><td><tt class="docutils literal">&quot;</tt></td>
<td>0.63 -- 0.79</td>
<td><tt class="docutils literal">1</tt></td>
<td>0.02  -- 0.025</td>
</tr>
<tr><td><tt class="docutils literal">#</tt></td>
<td>0.5  -- 0.63</td>
<td><tt class="docutils literal">2</tt></td>
<td>0.016 -- 0.02</td>
</tr>
<tr><td><tt class="docutils literal">$</tt></td>
<td>0.4  -- 0.5</td>
<td><tt class="docutils literal">3</tt></td>
<td>0.013 -- 0.016</td>
</tr>
<tr><td><tt class="docutils literal">%</tt></td>
<td>0.32 -- 0.4</td>
<td><tt class="docutils literal">4</tt></td>
<td>0.01  -- 0.013</td>
</tr>
<tr><td><tt class="docutils literal">&amp;</tt></td>
<td>0.25 -- 0.32</td>
<td><tt class="docutils literal">5</tt></td>
<td>0.0079 -- 0.01</td>
</tr>
<tr><td><tt class="docutils literal">'</tt></td>
<td>0.2  -- 0.25</td>
<td><tt class="docutils literal">6</tt></td>
<td>0.0063 -- 0.0079</td>
</tr>
<tr><td><tt class="docutils literal">(</tt></td>
<td>0.16 -- 0.2</td>
<td><tt class="docutils literal">7</tt></td>
<td>0.005  -- 0.0063</td>
</tr>
<tr><td><tt class="docutils literal">)</tt></td>
<td>0.13 -- 0.16</td>
<td><tt class="docutils literal">8</tt></td>
<td>0.004  -- 0.005</td>
</tr>
<tr><td><tt class="docutils literal">*</tt></td>
<td>0.1  -- 0.13</td>
<td><tt class="docutils literal">9</tt></td>
<td>0.0032 -- 0.004</td>
</tr>
<tr><td><tt class="docutils literal">+</tt></td>
<td>0.079 -- 0.1</td>
<td><tt class="docutils literal">:</tt></td>
<td>0.0025 -- 0.0032</td>
</tr>
<tr><td><tt class="docutils literal">,</tt></td>
<td>0.063 -- 0.079</td>
<td><tt class="docutils literal">;</tt></td>
<td>0.002  -- 0.0025</td>
</tr>
<tr><td><tt class="docutils literal">-</tt></td>
<td>0.05  -- 0.063</td>
<td><tt class="docutils literal">&lt;</tt></td>
<td>0.0016 -- 0.002</td>
</tr>
<tr><td><tt class="docutils literal">.</tt></td>
<td>0.04  -- 0.05</td>
<td><tt class="docutils literal">=</tt></td>
<td>0.0013 -- 0.0016</td>
</tr>
<tr><td><tt class="docutils literal">/</tt></td>
<td>0.032 -- 0.04</td>
<td><tt class="docutils literal">&gt;</tt></td>
<td>0.001  -- 0.0013</td>
</tr>
</tbody>
</table>
</blockquote>
<p>Note that each alignment is grown from a &quot;core&quot; region, and the
ambiguity estimates assume that the core is correctly aligned.  The
core is indicated by &quot;~&quot; symbols, and it contains exact matches only.</p>
<p>LAST has options to find alignments with optimal column probabilities,
instead of optimal score: see <a class="reference external" href="lastal.html">lastal.html</a>.</p>
</div>
</div>
</body>
</html>
